File: Programacion/S4/App_Punteros_CPP/Pila.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/* 
 * File:   Pila.cpp
 * Author: The Survivor
 * 
 * Created on 30 de marzo de 2020, 04:27 PM
 */
#include <iostream>
#include "Pila.h"

Pila::Pila()
{
    datos = new int[MAX];
    tope = -1;
    for(int i = 0; i < tam; i++)
        datos[i] = 0;//inicializa a ceros el arreglo
     //obligatorio inicializar en c++
     
    undato = datos;
    
     for(int i = 0; i<tam; i++, undato++){
         datos[i] = 0;
     }
}

bool Pila::push(int valor)
{
    if(tope == tam -1)
        return false;
    else
    {
        datos[++tope]= valor;
        return true;
    }
}

//pase de la referencia
bool Pila::pop(int &valor)
{
     if(tope == -1)
        return false;
     else
     {
        valor = datos[tope--];
        return true;
     }
}

void Pila::mostrar()
{
    cout<<"\nDatos en la pila : \n";
    for(int i = 0; i <= tope; i++)
        cout<<"\n Dato[ "<<i<<" ] -> "<<datos[i];
    
    cout<<endl;
    undato = &datos[0];
    for(int i = 0; i<tam; i++){
        cout<<"Dir de dato: " <<i <<" " <<&datos[i] <<endl;
        cout<<"Dir de undato + i"<<" " <<*(undato+i) <<endl;
    }
}

File: Programacion/S4/01_Apps_EstructuraDatos/App_NotacionPolaca_Pila/src/app_pila1/Pila.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package app_pila1;

/**
 *
 * @author user
 */
class Pila {
    int []vec;
    private boolean vacia;
    public boolean llena;
    int tope;
    
    public Pila(int n)
    {
        vec=new int[n];
        tope=0;
        vacia=true;
        llena=false;
        
    }
    
    public void push(int valor)
    {
        vacia=false;
        vec[tope++]=valor;
        if(tope==vec.length)
            llena=true;
    }
    
    public int pop()
            {
        if (--tope==0)
        vacia=true;
        llena=false;
        return vec[tope];
        
            }
    
    public boolean esta_vacia()
    {
       return vacia;
    }
    
    public boolean estaLlena()
    {
        return llena;
    }
    
    public int Tope(){
        return vec[tope-1];
    }
            
}
